### PR TITLE
Update `group by` parser changes to support Wildcard BP

### LINKED ISSUE #37740: [Bug]: Update `group by` parser changes to support Wildcard BP
### Description

$title

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by suleka96**
Fixes #37740

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by rdulmina — APPROVED** [review]

**Review by rdulmina — COMMENTED** [review]
We need to update the tests right?

File: misc/syntax-api-calls-gen/src/main/resources/api_gen_syntax_tree_descriptor.json [hunk 1]
```diff
@@ -4211,7 +4211,7 @@
         },
         {
           "name": "variableName",
```
Comment (lochana-chathura) [inline misc/syntax-api-calls-gen/src/main/resources/api_gen_syntax_tree_descriptor.json:4213]: We can call this `simpleBindingPattern` as `variableName` could be misleading? @rdulmina wdyt?
(In the AST we have `BLangSimpleBindingPattern` to represent both capture and wildcard binding patterns.)

Comment (rdulmina) [reply misc/syntax-api-calls-gen/src/main/resources/api_gen_syntax_tree_descriptor.json:4213]: +1

Comment (lochana-chathura) [reply misc/syntax-api-calls-gen/src/main/resources/api_gen_syntax_tree_descriptor.json:4213]: @suleka96 please note.

**Comment by suleka96** [comment]
The tests would't change cause previously, before changing it to `token` it was `capture binding pattern` so the context we had was `BINDING_PATTERN_STARTING_IDENTIFIER` since it applies to this change as well... error recovery stuff didn't change. will add more tests with wildcard bp.

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/QueryExpressionTest.java [hunk 2]
```diff
@@ -144,6 +144,31 @@ public void testQueryWithGroupByUsingMultipleGroupingKeys() {
...
 
+    @Test
+    public void testQueryWithGroupByUsingVarDefGroupingKeyWithWildcardBP1() {
+        testFile("query-expr/query_expr_source_110.bal", "query-expr/query_expr_assert_110.json");
+    }
+
+    @Test
+    public void testQueryWithGroupByUsingVarDefGroupingKeyWithWildcardBP2() {
+        testFile("query-expr/query_expr_source_111.bal", "query-expr/query_expr_assert_111.json");
+    }
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/QueryExpressionTest.java:155]:
```suggestion
    @Test
    public void testQueryWithGroupByUsingVarDefGroupingKeyWithWildcardBP() {
        testFile("query-expr/query_expr_source_110.bal", "query-expr/query_expr_assert_110.json");
        testFile("query-expr/query_expr_source_111.bal", "query-expr/query_expr_assert_111.json");
    }
```

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/QueryExpressionTest.java [hunk 3]
```diff
@@ -144,6 +144,31 @@ public void testQueryWithGroupByUsingMultipleGroupingKeys() {
...
+
+    @Test
+    public void testQueryWithGroupByUsingMultipleGroupingKeysWithWildcardBP2() {
+        testFile("query-expr/query_expr_source_113.bal", "query-expr/query_expr_assert_113.json");
+    }
+
+    @Test
+    public void testQueryWithGroupByUsingMultipleGroupingKeysWithWildcardBP3() {
+        testFile("query-expr/query_expr_source_114.bal", "query-expr/query_expr_assert_114.json");
+    }
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/QueryExpressionTest.java:170]:
```suggestion
    @Test
    public void testQueryWithGroupByUsingMultipleGroupingKeysWithWildcardBP() {
        testFile("query-expr/query_expr_source_112.bal", "query-expr/query_expr_assert_112.json");
        testFile("query-expr/query_expr_source_113.bal", "query-expr/query_expr_assert_113.json");
        testFile("query-expr/query_expr_source_114.bal", "query-expr/query_expr_assert_114.json");
    }
```

**Comment by lochana-chathura** [comment]
PR LGTM. @suleka96 you can skip the failing testings in the parserTestRunner mentioning the issue and enable them in PR #37555

**Comment by suleka96** [comment]
@lochana-chathura wouldn't it be better to keep it as it is than disabling and enabling again, since these are known failures (Due to missing classes in compiler side - `Node not supported: GroupByClauseNode` errors).  And also then we can easily see to that these are fixed in the semantic phase. Anyway the failing ones will immediately be addressed in the semantic phase.

**Comment by lochana-chathura** [comment]
I'm partially agreeing with you. However,
[1]. Depending on the build script all the tests may not be executed when we have test failures in the middle.
[2]. We have to manually go through the build log from top to bottom to verify what tests are failing and whether they are expected.

I think it is better if we can get the build pass. Merging a PR with a build failure is not a good practice IMO.

**Comment by rdulmina** [comment]
> I'm partially agreeing with you. However, [1]. Depending on the build script all the tests may not be executed when we have test failures in the middle. [2]. We have to manually go through the build log from top to bottom to verify what tests are failing and whether they are expected.
> 
> I think it is better if we can get the build pass. Merging a PR with a build failure is not a good practice IMO. @MaryamZi @gimantha WDYT?

+1

**Comment by gimantha** [comment]
> I'm partially agreeing with you. However, [1]. Depending on the build script all the tests may not be executed when we have test failures in the middle. [2]. We have to manually go through the build log from top to bottom to verify what tests are failing and whether they are expected.
> 
> I think it is better if we can get the build pass. Merging a PR with a build failure is not a good practice IMO. @MaryamZi @gimantha WDYT?

We can create a separate issue for enabling the tests and add that issue as a subtask in the main issue. 

- [ ]  Fix failing parser tests for query-expr

**Comment by suleka96** [comment]
Created #37888 to track allowing the failing tests

**Review by lochana-chathura — APPROVED** [review]

**Pull request merged**